### ISSUE TITLE
Migrate to a more actively maintained go memcached client

### ISSUE THREAD

**Issue by Andrewpk** [cortexproject/cortex#3980]
**Is your feature request related to a problem? Please describe.**
Seeing ingester errors such as `ingester-XX ingester ts=2021-03-19T21:09:31.02197084Z caller=spanlogger.go:53 traceID=XXX method=Memcache.GetMulti level=error msg="Failed to get keys from memcached" err="read tcp IP:60986->IP:11211: i/o timeout"`.
After debugging this issue, I'm down to the point where I'd have to start debugging the memcached client, at which point I noticed the [client cortex is using](https://github.com/bradfitz/gomemcache) is significantly unmaintained.

**Describe the solution you'd like**
Investigate and switch to a more actively maintained go memcache client.

**Comment by alvinlin123** [comment]
Would be nice that if we do want to move to a new memcached client, we also look into client that support AWS ElastiCache's [Auto Discovery](https://docs.aws.amazon.com/AmazonElastiCache/latest/mem-ug/AutoDiscovery.HowAutoDiscoveryWorks.html)

**Comment by pracucci** [comment]
> After debugging this issue, I'm down to the point where I'd have to start debugging the memcached client, at which point I noticed the client cortex is using is significantly unmaintained.

I understand the concerns about using an unmaintained client (even if we could always fork and maintain it) but what's the actual problem we want to solve? The `i/o timeout` doesn't look an issue caused by the specific client we're using.

> Would be nice that if we do want to move to a new memcached client, we also look into client that support AWS ElastiCache's Auto Discovery

Service discovery is something we should built on top of memcached client (we already have a service discovery mechanism and also sharding is built on top of it). I'm not sure this comment is a good argument in favour of changing library.

**Comment by bboreham** [comment]
We're actually using a fork of that library:
https://github.com/cortexproject/cortex/blob/dd9de3bcbfc0a46fcdf22fb76112fec96f775b44/go.mod#L86-L87

**Comment by alvinlin123** [comment]
> Service discovery is something we should built on top of memcached client (we already have a service discovery mechanism and also sharding is built on top of it). I'm not sure this comment is a good argument in favour of changing library.

I am not using this as an argument to push a change the client, but more like if there are interest to change, the AWS service discovery support would be a nice feature to have. Also, I am aware that Cortex has service discover on top of the memcached using DNS resolution. But the service discovery I have in mind is where ElastiCache has a feature where you can sed a `config get` memcached command to the config endpoint and get a list of all the cached nodes; we can build this feature on top of the memcached client, but I think it would make more sense for it to be supported in memcached client itself as AWS' service discovery feature depends on client itself. 

Also one thing I noticed about the memcached client currently being used by Cortex is doing a [modulo load balancing](https://github.com/themihai/gomemcache/blob/24332e2d58ababdccb318d731f4d149de21ae3e0/memcache/selector.go#L128), but this has the draw back whenever we scale the memcached node we can get a lot of cache miss; something like consistent hashing here might be more desirable.

**Comment by jtlisi** [comment]
I am open to switching to a new Memcache client or forking the code from the upstream library. However, I think it's important to be clear on why we would be making this change. For me personally I feel the following reasons are valid for updating the Memcache client library:

- **Poor instrumentation:** the upstream library is not currently instrumented with Prometheus (or any) metrics. However, our implementation wraps request with instrumentation so this is acceptable at the moment. However, having more granular debug logging directly from the memcached client could be very helpful when debugging some of these issues.
- **TLS:** Memcache [supports TLS](https://github.com/memcached/memcached/wiki/TLS). We should update our client to take advantage of this to enable Cortex to work in environments where you need to protect your traffic over wire.

If either of the two above points are addressed by another Memcache library or forking the library in this repository, I'd be open to the change.

**Comment by alvinlin123** [comment]
Maybe the what we are trying to achieve here is not merely to switch to a new memcached but more of to use a more actively maintained memcached client. So one of the options can still be to explore if we cam somehow have the current memcached client more actively maintained.

**Comment by bboreham** [comment]
> something like consistent hashing here might be more desirable.

Cortex uses consistent hashing by default for memcached. 
https://github.com/cortexproject/cortex/blob/8a2e2c1eeb65c2c987dc41dd920e3cec4728427d/pkg/chunk/cache/memcached_client.go#L89


 > whenever we scale the memcached node we can get a lot of cache miss

Possibly an effect of this:
https://github.com/cortexproject/cortex/blob/8a2e2c1eeb65c2c987dc41dd920e3cec4728427d/pkg/chunk/cache/memcached_client_selector.go#L55-L57

TL;DR "use a statefulset"; if you do not use Kubernetes then find some other way to arrange that names are stable and retain the same order after scaling.
You also want to use the `...memcached.addresses=dnssrvnoa+<DNSname>` form to ensure memcached-client is working with the DNS names (not IP addresses).

**Comment by alvinlin123** [comment]
> Cortex uses consistent hashing by default for memcached.

Cool it even use jump consistent hash, very cool!! Thanks for the pointer.
